datastream-kafka/src/main/java/com/linkedin/datastream/kafka/KafkaTransportProvider.java

```diff
@@ -136,6 +140,42 @@ private int getSourcePartitionFromEvent(BrooklinEnvelope event) {
         event.getMetadata().getOrDefault(BrooklinEnvelopeMetadataConstants.SOURCE_PARTITION, "-1"));
   }
 
+  @Override
+  public DatastreamRecordMetadata broadcast(String destinationUri, DatastreamProducerRecord record, SendCallback onEventComplete) {
+    Validate.isTrue(record.isBroadcastRecord(), "Trying to broadcast a non-broadcast type record.");
+
+    // Currently destination topic partition count will be queried from datastream destination metadata. This implies
+    // a restriction that datastream destination partition should be updated for broadcast to work correctly.
+    int partitionCount = _datastreamTask.getDatastreamDestination().getPartitions();
+    String topicName = KafkaTransportProviderUtils.getTopicName(destinationUri);
+
+    LOG.debug("Broadcasting record {} to all {} partitions of destination {}", record, partitionCount, destinationUri);
+    int partition = 0;
+    List<Integer> sentToPartitions = new ArrayList<>();
+    try {
+      for (; partition < partitionCount; partition++) {
+        record.setPartition(partition);
```

Comment:
I think modifying the record like this will get us into trouble. The caller would not expect the object to be modified, and the internal send() is probably expecting an immutable object. For example, what if send() were to buffer the record before looking at the partition?

I think we need to construct a new record for each send, either by shallow-copying into a new record (which is dangerous, cuz the fields to copy may change over time), or by wrapping the record in an outer class that overrides getPartition. Then you could do:

`send(new DatastreamProducerRecordWithPartititionOverride(record, newPartition))`

or smth.

Reply:
Went the copy route